### PR TITLE
[Model] Mamba2: Apply the gate inside the scan in batch-invariant mode

### PR DESCRIPTION
**Pull request by NolenLiang**
## Purpose

Stacked on #55905 (which is stacked on #57539, #55627 and #55895): only the last three commits are new. Tracked in #55524.

The goal of the stack is that a vLLM rollout reproduces the Megatron training forward of Mamba2 and hybrid models bit for bit. With #55627, #55895 and #55905 the SSD kernels of the two stacks agree once they run the same tiles. One difference remains inside the Mamba2 layer: where the gate is applied. Megatron's memory-efficient training path (the default, confirmed in NVIDIA/Megatron-LM#4871) multiplies the scan output by `silu(z)` in fp32 inside the chunk-scan kernel and then runs an ungated grouped RMSNorm. vLLM rounds the scan output first and applies the gate inside the gated RMSNorm. About 35% of the layer's output elements differ between the two orders.

In batch-invariant mode this PR follows the training order:

- The replayed prefill scan passes the gate into the varlen chunk-scan kernel (zeros for re-fed rows, whose outputs are discarded), and the row-gated emit kernel of #55905 multiplies the emitted row by `silu(z)` after the D residual, in the same order as the chunk-scan kernel.
- The mixer's RMSNorm runs ungated and, whenever every rank holds whole groups, uses the fp32 grouped Triton kernel for any group count; the torch path rounds differently. Under torch.compile the mixer norm custom op is off by default, so batch-invariant mode enables `+mixer2_gated_rms_norm` for models with Mamba layers and rejects an explicit `-mixer2_gated_rms_norm`.
- The SSD warmup also compiles the gated scan specialization.

Default mode is unchanged. In batch-invariant mode the layer is now numerically different from default mode on purpose: the mode exists to match the training forward.

Scope of the alignment claim: shapes where every tensor-parallel rank holds whole groups (`n_groups % tp_size == 0`, which is also the only configuration Megatron trains). With `n_groups == 1` and TP > 1 the norm keeps the torch fallback as an inference-only path.

Not a duplicate: no open PR changes the gate placement; #54993 keeps vLLM's order.

## Test Plan

```bash
VLLM_BATCH_INVARIANT=1 pytest tests/kernels/mamba/test_mamba2_ssd_emit.py tests/kernels/mamba/test_mamba2_exact_replay.py tests/kernels/mamba/test_mamba2_ssd_batch_invariance.py
pytest tests/kernels/mamba/test_mamba_mixer2.py -k "batch_invariant or fallback"
pytest tests/v1/attention/test_mamba2_batch_invariant.py
pytest tests/v1/determinism/test_mamba2_batch_invariance.py
```

Cross-stack check, out of tree: the vLLM batch-invariant SSD plus norm kernel composition (replayed prefill split at an unaligned point, token-by-token emit decode, ungated grouped norm) against Megatron's `mamba_chunk_scan_combined_varlen(z=...)` followed by the mamba_ssm ungated norm, same weights and inputs, granite-4.0-h and Nemotron-H-8B shapes, chunk 128 and 256, both sides on the pinned tiles. A limited model-level check on Nemotron-H-8B-Base (one 578-token prompt, GB200, bf16, TP1, eager prefill, this stack plus #52905, training side forced to the same cumsum head tile) finds the first Mamba layer's output and the residual stream after it bitwise identical to the Megatron training forward; the first divergence is the following MLP layer's RMSNorm.

Engine-level oracle, same harness as #55627: mamba2-130m (8 requests, with and without chunked prefill, CUDA graphs and eager), granite-4.0-h-350m under one preemption (CUDA graphs and eager), Nemotron-H-8B-Base-8K at TP 1, 2 and 4 with full CUDA graphs. The harness forces tokens through a custom logits processor, which makes the engine fall back to model runner V1; the in-tree determinism tests also pass with `VLLM_USE_V2_MODEL_RUNNER=1` (mamba2-130m, TP1).

## Test Result

GB200, Triton 3.7.1.

- 37 kernel-level cases pass (emit tests at chunk sizes 64, 128 and 256), including the gated variants against the varlen kernel with the gate inside; 6 norm cases pass (the module's forward reaches the fp32 grouped kernel for 1 and 8 groups, with and without a gate; the torch fallback keeps its variance in fp32); 11 metadata and config cases pass (the op is enabled for Mamba models only, an explicit `-mixer2_gated_rms_norm` is rejected).
- Both determinism tests pass with CUDA graphs, also with `VLLM_USE_V2_MODEL_RUNNER=1`.
- Kernel composition versus Megatron: bitwise identical scan rows and norm outputs for all four shape and chunk combinations when both sides use the pinned tiles; with autotuning on both sides the chunk-128 cases differ.
- Oracle: 0 mismatching logprobs in every run listed above, all on the final commit.
- pre-commit passes on the changed files.

## AI assistance

AI tools assisted with the implementation, tests and documentation. I reviewed every changed line and ran the tests above.

### CODE REVIEW

**Comment by NolenLiang** [comment]
Rebased onto the new #55905 head (57f727e03): the gate now goes through the simplified emit kernel from that PR, and the determinism tests pick up its explicit engine teardown. No functional change in this PR; kernel (18), norm (6), attention (13) and both determinism tests pass, and the Nemotron-H-8B TP=1/TP=2 CUDA-graph runs have zero mismatches on the new head.

**Comment by NolenLiang** [comment]
Rebased on the new #55905 head. No functional change. Tests in the description are from the new head.
